### ISSUE TITLE
[Tiny fix] Track mouse outside Window (on win32 impl)

### ISSUE THREAD

**Issue by lamogui** [LaurentGomila/SFML#764]
Hello
Here is a proposition, it's just a tiny fix i do every time there is a new SFML version and may be it's could be done directly on SFML code.
I use my own window class when I use sfml on window which have no border and resize triangle item to avoid interrupt opengl when resizing.

But on win32 impl there is a function that remove mouse tracking when the mouse is outside the client area...
It's problematic because when i resize and my mouse go outside the window i don't receive mouse move events and it 's not resizing the window anymore...

There is no reason why the win32 impl should not send outside window mouse event, because it's done with X11 implementation (at least it works on KDE without any fix).

If the user does'nt want use outside window mouse event he just have to catch the MouseLeft and MouseEnter events....

When i trash fix here; https://github.com/LaurentGomila/SFML/blob/master/src/SFML/Window/Win32/WindowImplWin32.cpp

I remove all setTracking calls inside void WindowImplWin32::processEvent(UINT message, WPARAM wParam, LPARAM lParam) then i had one in win32 impl constructors. Of course I push MouseMove event after the MouseLeft...

Another solution may to call them when the window gain or lost focus... but in any case continu to send MouseMove Events

**Comment by Mischa-Alff** [comment]
I don't get mouseMove events on ArchLinux x86_64 with X11 and XFCE4 (and XFWM) when the mouse has exited the window.

Here's my test code:
```c++
#include <iostream>
#include <SFML/Window.hpp>
int main() {
        sf::Window window{{800,600}, "test"};
        while(window.isOpen()) {
                sf::Event event;
                while(window.pollEvent(event)) {
                        if(event.type==sf::Event::Closed)
                                window.close();
                        else if(event.type==sf::Event::MouseMoved)
                                std::cout<<"Mouse Moved."<<std::endl;
                }
        }
}
```

**Comment by eXpl0it3r** [comment]
Thanks @Mischa-Alff for testing.

@lamogui What exactly do you change, because I don't [see any checks](https://github.com/LaurentGomila/SFML/blob/master/src/SFML/Window/Win32/WindowImplWin32.cpp#L854) for the Mouse Move event. If the `WM_MOUSEMOVE` message is being received then you'll get a Mouse Move event.
But maybe I just overlooked it or didn't fully understand your problem.

Have you thought about using the real-time input `sf::Mouse::getPosition()`?

**Comment by Mischa-Alff** [comment]
Not sure about Windows, but on Linux, `sf::Keyboard::isKeyPressed` is extremely slow, and `sf::Mouse::getPosition` isn't too bad, since it's usually called once per frame, but it's not exactly fast either.

**Comment by eXpl0it3r** [comment]
True, however it would still be sufficient during the "resizing" you don't have to make the call while window isn't being resized.

**Comment by MarioLiebisch** [comment]
> There is no reason why the win32 impl should not send outside window mouse event, because it's done with X11 implementation (at least it works on KDE without any fix).

There is significant reason, because that's how the underlying window system (and basically all UI toolkits I know) handle mouse movement. You don't get any mouse move events if the cursor is outside your client area.

Even in HTML you won't get mouse move events unless the cursor is above the related element.

Right now it's far more likely that users assume that there won't be any mouse move event unless the cursor is inside the SFML window.

> I use my own window class when I use sfml on window which have no border and resize triangle item to avoid interrupt opengl when resizing.

That's something you have to handle in your own window code, not inside SFML.

---

Also make sure you're using the latest version of SFML, since you should receive further mouse move events in case the cursor is dragged outside while a button is pushed.

**Comment by eXpl0it3r** [comment]
Kind of lost the overview here, but is still now actually a "feature"/bugfix we want or need?

**Comment by lamogui** [comment]
nevermind. Like I previously write it seems fixed on win32 since 2.2. I don't know on linux. This issue should be closed.
